src/2024h2/Contracts-and-invariants.md

```diff
@@ -0,0 +1,220 @@
+
+But in my shiny future, people get to reuse the majority of their contracts and just "plug in" different dynamic-validation and static-verification tools, and all the tools know how to leverage the common contract language that is built into Rust.
+
+## Design axioms
+
+[da]: ../about/design_axioms.md
+
+Felix [presented these axioms][rw2024] as "Shared Values" at the 2024 Rust Verification Workshop.
+
+[rw2024]: http://pnkfx.org/presentations/contracts2-rw2024.pdf
+
+### 1. Purpose: Specification Mechanism first, Verification Mechanism second
+
+Contracts have proven useful under the "Design by Contract" philosophy, which usually focuses on pre + post + frame conditions (also known as "requires", "ensures", and "modifies" clauses in systems such as the Java Modelling Language). In other words, attaching predicates to the API boundaries of code, which makes contracts a *specification mechanism*.
+
+There are other potential uses for attaching predicates to points in the code, largely for encoding formal correctness arguments. My main examples of these are Representation Invariants, Loop Invariants, and Termination Measures (aka "decreasing functions").
+
+In an ideal world, contracts would be useful for both purposes. But if we have to make choices about what to prioritize, we should focus on the things that make contracts useful as an API specification mechanism. In my opinion, API specification is the use-case that is going to benefit the broadest set of developers.
+
+### 2. Contracts should be (semi-)useful out-of-the-box
+
+This has two parts: 
+
+Anyone can eat: I want any Rust developer to be able to turn on "contract checking" in some form without having to change toolchain nor install 3rd-party tool, and get *some* utility from the result. (It's entirely possible that contracts become *even more* useful when used in concert with a suitable 3rd-party tool; that's a separate matter.)
+
+Anyone can cook: Any Rust developer can also *add contracts* to their own code, without having to change their toolchain.
+
+### 3. Contracts are not just assertions
+
+Contracts are meant to enable modular reasoning.
+
+Contracts have both a dynamic semantics and a static semantics.
+
+In an ideal dynamic semantics, a broken contract will identify *which component* is at fault for breaking the contract. (We do acknowledge that precise blame assignment becomes non-trivial with dynamic dispatch.)
+
+In an ideal static semantics, contracts enable theorem provers to choose, instead of reasoning about `F(G)`, to instead allow independent correctness proofs for `F(...)` and `... G ...`.
```

Comment:
I'd really like concrete examples here, I can't follow this section. In what way do these arguments support "contracts are not just assertions"? Couldn't simple propositional pre/post-conditions (i.e. assertions) have the properties you list there?

Reply:
Imagine a function `mul2` where the contract never actually says that it multiplies by two, but other people's code calling this function still relies on this not-contractually-specified behavior.

Dynamic checking can never find this fact since no execution ever causes a big, since the dynamic values always "happen to" also uphold a post condition that you never added to the formal spec. But symbolic reasoning tools will consider each function abstractly, and then notice that the remaining code is wrong since there also odd values are considered since the post condition never explicitly excluded them. Contracts are meant to enforce such modular reasoning, even if some tools would not actually check for this.

So it's about what a postcondition means: It's an assertion run when a method returns, but _also_ an abstraction barrier.

At least that's how I read this point.

Reply:
@Nadrieril I'm not sure if you're asking about clarification regarding the dynamic semantics or static semantics. The thrust of this point is largely about the dynamic semantics.

My most immediate example of where "contracts are not just assertions" is a case like this ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3b3b8e3ba3cb1211229fc1318cd3a384)):

```rust
// requires: (nothing)
fn apply_to_three(f: impl Fn(u32)) { f(3); }

// requires: x is even
fn i_only_like_even_numbers(x: u32) { assert!(x % 2 == 0); }

fn i_should_be_blamed_for_an_invalid_composition() -> impl Fn() {
    || apply_to_three(i_only_like_even_numbers)
}

fn main() {
    let g = i_should_be_blamed_for_an_invalid_composition();
    g();
}
```

the function `i_should_be_blamed_for_an_invalid_composition` is the component at fault: It passed a function with a specified precondition into a function that put no preconditions on its input (and thus implicitly claimed it would handle any function at all).

Systems like Racket, I believe, properly assign blame in this case. (I don't have time at the moment to translate the above into a Racket example, but I'll look into doing so later.)

Reply:
(heh. I used a level of indirection because I wanted to ensure the stack trace wouldn't actually include the component at fault. I *totally* forgot that the closure would include the name of that component in its mangled symbol...)

Reply:
I see, thanks for the explanations. Felix, is this blame tracking something rustc could do eventually or will it need external tooling?

Reply:
To be honest, I'm not sure it can be done efficiently, in full generality, via rustc alone. The typical implementation as shown in Racket approach entails adding wrappers around closures, and I think the Racket folks have found that turning on all contract checking leads to unacceptable overhead (but this is in part anecdotal).